File: problems/problems_1802/solution.py
import solution


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.maxValue(*test_input)

    def maxValue(self, n, index, maxSum):
        """
        :type n: int
        :type index: int
        :type maxSum: int
        :rtype: int
        """
        def helper(val):
            # sum = (start+end) * len // 2
            if val > index:
                l = (val - 1 + val - index) * index // 2
            else:
                l = (val - 1 + 1) * (val - 1) // 2 + (index - val + 1)
            if val > n - 1 - index:
                r = (val - 1 + val - n + 1 + index) * (n - 1 - index) // 2
            else:
                r = (val - 1 + 1) * (val - 1) // 2 + (n - 1 - index - val + 1)
            return l + r > maxSum - val

        left, right = 1, maxSum - n + 1
        while left < right:
            mid = (left + right) // 2
            if helper(mid):
                right = mid
            else:
                left = mid + 1

        return left if not helper(left) else left - 1

File: problems/problems_1802/Solution.cpp
//go:build ignore
#include "cpp/common/Solution.h"

using namespace std;
using json = nlohmann::json;

class Solution {
public:
  int maxValue(int n, int index, int maxSum) {
    int left = 1, right = maxSum - n + 1;
    auto check = [n, index, maxSum](int mid) {
      int64_t leftSum = 0, rightSum = 0;
      // 从0到index-1
      if (mid > index) {
        leftSum = (int64_t)(mid - 1 + mid - index) * index / 2;
      } else {
        leftSum = (int64_t)(mid - 1 + 1) * (mid - 1) / 2 + (index - mid + 1);
      }
      // 从index+1到n-1
      if (mid > n - index - 1) {
        rightSum =
            (int64_t)(mid - 1 + mid - (n - index - 1)) * (n - index - 1) / 2;
      } else {
        rightSum = (int64_t)(mid - 1 + 1) * (mid - 1) / 2 +
                   ((n - index - 1) - mid + 1);
      }
      return leftSum + rightSum <= maxSum - mid;
    };
    while (left < right) {
      int mid = left + (right - left + 1) / 2;
      if (check(mid)) {
        left = mid;
      } else {
        right = mid - 1;
      }
    }
    return left;
  }
};

json leetcode::qubh::Solve(string input_json_values) {
  vector<string> inputArray;
  size_t pos = input_json_values.find('\n');
  while (pos != string::npos) {
    inputArray.push_back(input_json_values.substr(0, pos));
    input_json_values = input_json_values.substr(pos + 1);
    pos = input_json_values.find('\n');
  }
  inputArray.push_back(input_json_values);

  Solution solution;
  int n = json::parse(inputArray.at(0));
  int index = json::parse(inputArray.at(1));
  int maxSum = json::parse(inputArray.at(2));
  return solution.maxValue(n, index, maxSum);
}
